### ISSUE TITLE
[PATCH] Deprecate setting ${^ENCODING}

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14240]
Migrated from [rt.perl.org#123210](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123210) (status was 'resolved')

Searchable as RT123210$

**Comment by p5pRT** [comment]
### From @ilmari

#### Created by @ilmari



See attached patch

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low
    Type=Patch
    PatchStatus=HasPatch

Site configuration information for perl 5.20.1:

Configured by Debian Project at Sun Oct 19 23:39:33 UTC 2014.

Summary of my perl5 (revision 5 version 20 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=3.2.0-4-amd64, archname=x86_64-linux-gnu-thread-multi
    uname='linux binet 3.2.0-4-amd64 #1 smp debian 3.2.63-2 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -D_FORTIFY_SOURCE=2 -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Dldflags= -Wl,-z,relro -Dlddlflags=-shared -Wl,-z,relro -Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.20 -Darchlib=/usr/lib/x86_64-linux-gnu/perl/5.20 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/x86_64-linux-gnu/perl5/5.20 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.20.1 -Dsitearch=/usr/local/lib/x86_64-linux-gnu/perl/5.20.1 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -Uversiononly -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.20.1 -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.9.1', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/4.9/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=libc-2.19.so, so=so, useshrplib=true, libperl=libperl.so.5.20
    gnulibc_version='2.19'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib -fstack-protector'

Locally applied patches:
    DEBPKG:debian/cpan_definstalldirs - Provide a sensible INSTALLDIRS default for modules installed from CPAN.
    DEBPKG:debian/db_file_ver - http://bugs.debian.org/340047 Remove overly restrictive DB_File version check.
    DEBPKG:debian/doc_info - Replace generic man(1) instructions with Debian-specific information.
    DEBPKG:debian/enc2xs_inc - http://bugs.debian.org/290336 Tweak enc2xs to follow symlinks and ignore missing @INC directories.
    DEBPKG:debian/errno_ver - http://bugs.debian.org/343351 Remove Errno version check due to upgrade problems with long-running processes.
    DEBPKG:debian/libperl_embed_doc - http://bugs.debian.org/186778 Note that libperl-dev package is required for embedded linking
    DEBPKG:fixes/respect_umask - Respect umask during installation
    DEBPKG:debian/writable_site_dirs - Set umask approproately for site install directories
    DEBPKG:debian/extutils_set_libperl_path - EU:MM: set location of libperl.a under /usr/lib
    DEBPKG:debian/no_packlist_perllocal - Don't install .packlist or perllocal.pod for perl or vendor
    DEBPKG:debian/prefix_changes - Fiddle with *PREFIX and variables written to the makefile
    DEBPKG:debian/fakeroot - Postpone LD_LIBRARY_PATH evaluation to the binary targets.
    DEBPKG:debian/instmodsh_doc - Debian policy doesn't install .packlist files for core or vendor.
    DEBPKG:debian/ld_run_path - Remove standard libs from LD_RUN_PATH as per Debian policy.
    DEBPKG:debian/libnet_config_path - Set location of libnet.cfg to /etc/perl/Net as /usr may not be writable.
    DEBPKG:debian/mod_paths - Tweak @INC ordering for Debian
    DEBPKG:debian/module_build_man_extensions - http://bugs.debian.org/479460 Adjust Module::Build manual page extensions for the Debian Perl policy
    DEBPKG:debian/prune_libs - http://bugs.debian.org/128355 Prune the list of libraries wanted to what we actually need.
    DEBPKG:fixes/net_smtp_docs - [rt.cpan.org #36038] http://bugs.debian.org/100195 Document the Net::SMTP 'Port' option
    DEBPKG:debian/perlivp - http://bugs.debian.org/510895 Make perlivp skip include directories in /usr/local
    DEBPKG:debian/deprecate-with-apt - http://bugs.debian.org/747628 Point users to Debian packages of deprecated core modules
    DEBPKG:debian/squelch-locale-warnings - http://bugs.debian.org/508764 Squelch locale warnings in Debian package maintainer scripts
    DEBPKG:debian/skip-upstream-git-tests - Skip tests specific to the upstream Git repository
    DEBPKG:debian/patchlevel - http://bugs.debian.org/567489 List packaged patches for 5.20.1-2 in patchlevel.h
    DEBPKG:debian/skip-kfreebsd-crash - http://bugs.debian.org/628493 [perl #96272] Skip a crashing test case in t/op/threads.t on GNU/kFreeBSD
    DEBPKG:fixes/document_makemaker_ccflags - http://bugs.debian.org/628522 [rt.cpan.org #68613] Document that CCFLAGS should include $Config{ccflags}
    DEBPKG:debian/find_html2text - http://bugs.debian.org/640479 Configure CPAN::Distribution with correct name of html2text
    DEBPKG:debian/perl5db-x-terminal-emulator.patch - http://bugs.debian.org/668490 Invoke x-terminal-emulator rather than xterm in perl5db.pl
    DEBPKG:debian/cpan-missing-site-dirs - http://bugs.debian.org/688842 Fix CPAN::FirstTime defaults with nonexisting site dirs if a parent is writable
    DEBPKG:fixes/memoize_storable_nstore - [rt.cpan.org #77790] http://bugs.debian.org/587650 Memoize::Storable: respect 'nstore' option not respected
    DEBPKG:debian/regen-skip - Skip a regeneration check in unrelated git repositories
    DEBPKG:fixes/regcomp-mips-optim - [perl #122817] http://bugs.debian.org/754054 Downgrade the optimization of regcomp.c on mips and mipsel due to a gcc-4.9 bug
    DEBPKG:debian/makemaker-pasthru - http://bugs.debian.org/758471 Pass LD settings through to subdirectories
    DEBPKG:fixes/perldoc-less-R - [rt.cpan.org #98636] http://bugs.debian.org/758689 Tell the 'less' pager to allow terminal escape sequences
    DEBPKG:fixes/pod_man_reproducible_date - http://bugs.debian.org/759405 Support POD_MAN_DATE in Pod::Man for the left-hand footer
    DEBPKG:fixes/data_dump_infinite_recurse - [19be3be] don't recurse infinitely in Data::Dumper
    DEBPKG:fixes/io_uncompress_gunzip_inmemory - http://bugs.debian.org/747363 [rt.cpan.org #95494] Fix gunzip to in-memory file handle
    DEBPKG:fixes/socket_test_recv_fix - http://bugs.debian.org/758718 [perl #122657] Compare recv return value to peername in socket test
    DEBPKG:fixes/hurd_socket_recv_todo - http://bugs.debian.org/758718 [perl #122657] TODO checking the result of recv() on hurd


@INC for perl 5.20.1:
    /home/ilmari/perl5/lib/perl5/x86_64-linux-gnu-thread-multi
    /home/ilmari/perl5/lib/perl5
    /etc/perl
    /usr/local/lib/x86_64-linux-gnu/perl/5.20.1
    /usr/local/share/perl/5.20.1
    /usr/lib/x86_64-linux-gnu/perl5/5.20
    /usr/share/perl5
    /usr/lib/x86_64-linux-gnu/perl/5.20
    /usr/share/perl/5.20
    /usr/local/lib/site_perl
    .


Environment for perl 5.20.1:
    HOME=/home/ilmari
    LANG=en_GB.utf8
    LANGUAGE=en_GB:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ilmari/perl5/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
    PERL5LIB=/home/ilmari/perl5/lib/perl5
    PERL_AUTOINSTALL_PREFER_CPAN=1
    PERL_BADLANG (unset)
    PERL_LOCAL_LIB_ROOT=/home/ilmari/perl5
    PERL_MB_OPT=--install_base "/home/ilmari/perl5"
    PERL_MM_OPT=INSTALL_BASE=/home/ilmari/perl5
    PERL_MM_PREFER_CPAN=1
    PERL_MM_USE_DEFAULT=1
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @ilmari

<details><summary>0001-Deprecate-setting-ENCODING.patch</summary>

```diff
From 095536837e8406ac88e02ee978057845ac57629a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Dagfinn=20Ilmari=20Manns=C3=A5ker?= <ilmari@ilmari.org>
Date: Mon, 10 Nov 2014 23:54:46 +0000
Subject: [PATCH 1/2] Deprecate setting ${^ENCODING}

---
 mg.c               |  4 ++++
 pod/perldiag.pod   |  5 +++++
 pod/perlvar.pod    |  4 ++--
 t/lib/warnings/mg  | 13 +++++++++++++
 t/re/pat_re_eval.t |  2 ++
 t/uni/chomp.t      |  2 ++
 6 files changed, 28 insertions(+), 2 deletions(-)

diff --git a/mg.c b/mg.c
index 4524d01..b36a828 100644
--- a/mg.c
+++ b/mg.c
@@ -2604,6 +2604,10 @@ Perl_magic_set(pTHX_ SV *sv, MAGIC *mg)
 	else if (strEQ(mg->mg_ptr+1, "NCODING")) {
 	    SvREFCNT_dec(PL_encoding);
 	    if (SvOK(sv) || SvGMAGICAL(sv)) {
+		if (PL_localizing != 2) {
+		    Perl_ck_warner_d(aTHX_ packWARN(WARN_DEPRECATED),
+				     "Setting ${^ENCODING} is deprecated");
+		}
 		PL_encoding = newSVsv(sv);
 	    }
 	    else {
diff --git a/pod/perldiag.pod b/pod/perldiag.pod
index 86a525b..4d9ca23 100644
--- a/pod/perldiag.pod
+++ b/pod/perldiag.pod
@@ -5161,6 +5161,11 @@ a positive integer, where the integer was the address of the reference.
 As of Perl 5.20.0 this is a fatal error, to allow future versions of Perl
 to use non-integer refs for more interesting purposes.
 
+=item Setting ${^ENCODING} is deprecated
+
+(W deprecated) You assiged a non-C<undef> value to C<${^ENCODING}>.
+This is deprecated for the same reasons L<encoding> is deprecated.
+
 =item shift on reference is experimental
 
 (S experimental::autoderef) C<shift> with a scalar argument is experimental
diff --git a/pod/perlvar.pod b/pod/perlvar.pod
index fe44c9d..61a2fdf 100644
--- a/pod/perlvar.pod
+++ b/pod/perlvar.pod
@@ -1873,8 +1873,8 @@ X<${^ENCODING}>
 
 The I<object reference> to the C<Encode> object that is used to convert
 the source code to Unicode.  Thanks to this variable your Perl script
-does not have to be written in UTF-8.  Default is I<undef>.  The direct
-manipulation of this variable is highly discouraged.
+does not have to be written in UTF-8.  Default is I<undef>.  Setting
+this variable to any other value is deprecated.
 
 This variable was added in Perl 5.8.2.
 
diff --git a/t/lib/warnings/mg b/t/lib/warnings/mg
index 7eb8428..3655a58 100644
--- a/t/lib/warnings/mg
+++ b/t/lib/warnings/mg
@@ -83,3 +83,16 @@ Use of uninitialized value $3 in oct at - line 3.
 use warnings 'uninitialized';
 $ENV{FOO} = undef; # should not warn
 EXPECT
+########
+${^ENCODING} = 42;
+{ local ${^ENCODING}; }
+${^ENCODING} = undef;
+{ local ${^ENCODING} = 37; }
+no warnings 'deprecated';
+${^ENCODING} = 42;
+{ local ${^ENCODING}; }
+${^ENCODING} = undef;
+{ local ${^ENCODING} = 37; }
+EXPECT
+Setting ${^ENCODING} is deprecated at - line 1.
+Setting ${^ENCODING} is deprecated at - line 4.
diff --git a/t/re/pat_re_eval.t b/t/re/pat_re_eval.t
index c77439a..16ecf35 100644
--- a/t/re/pat_re_eval.t
+++ b/t/re/pat_re_eval.t
@@ -744,7 +744,9 @@ sub run_tests {
 			# Poor man's "use encoding 'ascii'".
 			# This causes a different code path in S_const_str()
 			# to be used
+			no warnings 'deprecated';
 			local ${^ENCODING} = $enc;
+			use warnings 'deprecated';
 			use re 'eval';
 			ok($ss =~ /^$cc/, fmt("encode       $u->[2]", $ss, $cc));
 		    }
diff --git a/t/uni/chomp.t b/t/uni/chomp.t
index 4730736..c78c3e1 100644
--- a/t/uni/chomp.t
+++ b/t/uni/chomp.t
@@ -33,7 +33,9 @@ our %mbchars = (
 plan tests => 2 * (4 ** 3 + 4 + 1) * (keys %mbchars);
 
 for my $enc (sort keys %mbchars) {
+    no warnings 'deprecated';
     local ${^ENCODING} = find_encoding($enc);
+    use warnings 'deprecated';
     my @char = (sort(keys   %{ $mbchars{$enc} }),
 		sort(values %{ $mbchars{$enc} }));
 
-- 
2.1.1


```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

On 11/14/2014 03&#8203;:51 PM\, Dagfinn Ilmari Mannsåker \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Dagfinn Ilmari Mannsåker
> \# Please include the string&#8203;:  \[perl \#123210\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123210 >
>
>
> This is a bug report for perl from ilmari@&#8203;ilmari\.org\,
> generated with the help of perlbug 1\.40 running under perl 5\.20\.1\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> See attached patch
>

Thanks\, I've put it out to smoke\.  I've gotten very wary of doing C 
changes without at least a Win32 result\.  I'll apply it when the smokes 
look good\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

On 11/14/2014 04&#8203;:17 PM\, Karl Williamson wrote&#8203;:
> On 11/14/2014 03&#8203;:51 PM\, Dagfinn Ilmari Mannsåker \(via RT\) wrote&#8203;:
>> \# New Ticket Created by  Dagfinn Ilmari Mannsåker
>> \# Please include the string&#8203;:  \[perl \#123210\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123210 >
>>
>>
>> This is a bug report for perl from ilmari@&#8203;ilmari\.org\,
>> generated with the help of perlbug 1\.40 running under perl 5\.20\.1\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please describe your issue here\]
>>
>> See attached patch
>>
>
> Thanks\, I've put it out to smoke\.  I've gotten very wary of doing C
> changes without at least a Win32 result\.  I'll apply it when the smokes
> look good\.
>

Actually\, we can't push this to blead before the Encode piece is done\. 
We can do that temporarily\, if we choose\, in blead\, while waiting for 
the upstream to be applied\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Nov 14 14&#8203;:51&#8203;:00 2014\, ilmari wrote&#8203;:
> This is a bug report for perl from ilmari@&#8203;ilmari\.org\,
> generated with the help of perlbug 1\.40 running under perl 5\.20\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> See attached patch

Thank you\.  The perldiag entry needs D for the severity\, not W\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @ilmari

Karl Williamson \<public@&#8203;khwilliamson\.com> writes&#8203;:

> On 11/14/2014 04&#8203;:17 PM\, Karl Williamson wrote&#8203;:
>
> Actually\, we can't push this to blead before the Encode piece is
> done\. We can do that temporarily\, if we choose\, in blead\, while waiting
> for the upstream to be applied\.

FYI\, the Encode patch is at https://rt.cpan.org/Ticket/Display.html?id=100347


\-\- 
"A disappointingly low fraction of the human race is\,
 at any given time\, on fire\." \- Stig Sandbeck Mathisen

**Comment by p5pRT** [comment]
### From @ilmari

"Father Chrysostomos via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> Thank you\.  The perldiag entry needs D for the severity\, not W\.

Indeed it does\. I was confused by the "Setting $/ to a reference to %s"
entry\, above it\, which has \(W deprecated\)\.  The whole perldiag entry
could do with rewording\, since it currently references L\<encoding>\,
which says that the internal mechanism \(i\.e\. $\{^ENCODING\}\) is
deprecated\, which is a bit circular\.  I'll try and come up with a better
explanation\, but if someone beats me to it I won't mind\.

\-\- 
"I use RMS as a guide in the same way that a boat captain would use
 a lighthouse\.  It's good to know where it is\, but you generally
 don't want to find yourself in the same spot\." \- Tollef Fog Heen

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Nov 14 16&#8203;:03&#8203;:55 2014\, ilmari wrote&#8203;:
> "Father Chrysostomos via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> 
> > Thank you\.  The perldiag entry needs D for the severity\, not W\.
> 
> Indeed it does\. I was confused by the "Setting $/ to a reference to %s"
> entry\, above it\, which has \(W deprecated\)\.

Thank you for drawing my attention to it\.  Why\, if all other deprecation warnings are default warnings\, is this not one also?  I have corrected it in ddc3d23f0\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks\, applied \(with minor adjustments\) by 
d5b4785c1592d741608f5cff697058cc284db4eb
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
